File: dart/test/schema_conformance_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';

import 'package:modular_api/src/core/usecase/usecase.dart';
import 'package:modular_api/src/core/schema/field.dart';

class HelloInput extends Input {
  @Field(description: 'Name to greet')
  final String name;
  HelloInput({required this.name});

  factory HelloInput.fromJson(Map<String, dynamic> json) =>
      HelloInput(name: (json['name'] ?? '').toString());

  @override
  Map<String, dynamic> toJson() => {'name': name};

  @override
  List<SchemaField> get schemaFields => [
        SchemaField.string('name', description: 'Name to greet'),
      ];
}

class HelloOutput extends Output {
  @Field(description: 'Greeting message')
  final String message;
  HelloOutput({this.message = ''});

  @override
  int get statusCode => 200;

  @override
  Map<String, dynamic> toJson() => {'message': message};

  @override
  List<SchemaField> get schemaFields => [
        SchemaField.string('message', description: 'Greeting message'),
      ];
}

/// Loads a JSON fixture relative to the monorepo root.
Map<String, dynamic> loadFixture(String name) {
  // From dart/test/ → ../../tests/fixtures/
  final path = '${Directory.current.path}/../tests/fixtures/$name';
  final file = File(path);
  if (!file.existsSync()) {
    // When running from dart/ directory
    final altPath = '${Directory.current.path}/tests/fixtures/$name';
    final altFile = File(altPath);
    if (altFile.existsSync()) {
      return jsonDecode(altFile.readAsStringSync()) as Map<String, dynamic>;
    }
    throw FileSystemException('Fixture not found', path);
  }
  return jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
}

void main() {
  group('Schema Conformance', () {
    test('HelloInput schema matches shared fixture', () {
      final fixture = loadFixture('hello_input_schema.json');
      final input = HelloInput(name: 'test');
      expect(input.toSchema(), equals(fixture));
    });

    test('HelloOutput schema matches shared fixture', () {
      final fixture = loadFixture('hello_output_schema.json');
      final output = HelloOutput(message: 'test');
      expect(output.toSchema(), equals(fixture));
    });
  });
}

File: ts/test/schema_conformance.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { resolve } from 'path';

// Input/Output with @Field decorators — the new auto-schema pattern

import { Input, Output } from '../src/core/usecase';
import { Field } from '../src/core/schema/field';

class HelloInput extends Input {
  @Field.string({ description: 'Name to greet', example: 'World' })
  name!: string;

  static fromJson(json: Record<string, unknown>): HelloInput {
    const instance = new HelloInput();
    instance.name = (json['name'] ?? '').toString();
    return instance;
  }
}

class HelloOutput extends Output {
  @Field.string({ description: 'Greeting message', example: 'Hello, World!' })
  message!: string;

  get statusCode() {
    return 200;
  }

  static fromJson(json: Record<string, unknown>): HelloOutput {
    const instance = new HelloOutput();
    instance.message = (json['message'] ?? '').toString();
    return instance;
  }
}

const fixturesDir = resolve(__dirname, '../..', 'tests', 'fixtures');

function loadFixture(name: string): Record<string, unknown> {
  return JSON.parse(readFileSync(resolve(fixturesDir, name), 'utf-8'));
}

describe('Schema Conformance', () => {
  it('HelloInput schema matches shared fixture', () => {
    const fixture = loadFixture('hello_input_schema.json');
    const input = new HelloInput();
    expect(input.toSchema()).toEqual(fixture);
  });

  it('HelloOutput schema matches shared fixture', () => {
    const fixture = loadFixture('hello_output_schema.json');
    const output = new HelloOutput();
    expect(output.toSchema()).toEqual(fixture);
  });
});
